File: ptmipf/webui/static/examples.js
/* The examples page: build a polycrystal, fetch its potential, write the run.
 *
 * Everything expensive happens on the server; this side only collects the
 * settings, shows what a build would cost before it is asked for, and renders
 * what came back.
 */
"use strict";

const $ = (id) => document.getElementById(id);
const state = { meta: null, building: false };

async function api(path, options) {
  const response = await fetch(path, options);
  const payload = await response.json().catch(() => ({}));
  if (!response.ok) throw new Error(payload.error || `the server answered ${response.status}`);
  return payload;
}

function setStatus(text, kind) {
  const el = $("status");
  el.textContent = text;
  el.className = "status" + (kind ? " " + kind : "");
}

function settings() {
  return {
    box: parseFloat($("box").value),
    n_grains: parseInt($("grains").value, 10),
    strain: parseFloat($("strain").value),
    strain_rate: parseFloat($("strain-rate").value),
    temperature: parseFloat($("temperature").value),
    seed: parseInt($("seed").value, 10),
    builder: $("builder").value,
  };
}

/* The same arithmetic the server uses, so a card can quote a cost before
 * anything is built.  Both are estimates; the one after the build is the one
 * with the real atom count in it. */
const RATE_FOUR_CORES = 2.0e6;   // atom-steps a second, measured, see lammps.py
const RATE_ONE_CORE = 6.0e5;

function estimate(entry) {
  const s = settings();
  const atoms = Math.round(0.97 * entry.atoms_per_cell * Math.pow(s.box / entry.a0, 3));
  const steps = Math.round(s.strain / s.strain_rate / 0.002) + 2000;
  const atomSteps = atoms * steps;
  return {
    atoms,
    atomSteps,
    low: atomSteps / RATE_FOUR_CORES / 60,
    high: atomSteps / RATE_ONE_CORE / 60,
  };
}

function renderCatalogue() {
  const box = $("catalogue");
  box.replaceChildren();
  for (const entry of state.meta.examples) {
    const cost = estimate(entry);
    const card = document.createElement("div");
    card.className = "example";

    const title = document.createElement("h3");
    title.textContent = `${entry.element} polycrystal, ${entry.structure}`;
    const cite = document.createElement("p");
    cite.className = "muted note";
    const link = document.createElement("a");
    link.href = entry.url;
    link.target = "_blank";
    link.rel = "noreferrer";
    link.textContent = entry.citation;
    cite.append("Potential: ", link);

    const cost_line = document.createElement("p");
    cost_line.className = "note";
    cost_line.textContent =
      `about ${cost.atoms.toLocaleString()} atoms, ` +
      `${(cost.atomSteps / 1e6).toFixed(0)} million atom-steps: ` +
      `roughly ${cost.low.toFixed(0)} to ${cost.high.toFixed(0)} minutes to run`;

    const button = document.createElement("button");
    button.className = "primary";
    button.textContent = `Build ${entry.element}`;
    button.addEventListener("click", () => build(entry));

    card.append(title, cite, cost_line, button);
    box.append(card);
  }
}

async function build(entry) {
  if (state.building) return;
  state.building = true;
  setStatus(`building the ${entry.element} example, downloading its potential…`, "busy");
  for (const button of document.querySelectorAll(".example button")) button.disabled = true;
  try {
    const report = await api("/api/examples/build", {
      method: "POST",
      headers: { "Content-Type": "application/json" },
      body: JSON.stringify({ element: entry.element, ...settings() }),
    });
    renderResult(report);
    setStatus(`built ${report.n_atoms.toLocaleString()} atoms in ${report.relative}`, "");
  } catch (error) {
    setStatus(error.message, "error");
    renderFailure(error.message);
  } finally {
    state.building = false;
    for (const button of document.querySelectorAll(".example button")) button.disabled = false;
  }
}

function renderFailure(message) {
  $("result-card").hidden = false;
  const box = $("result");
  box.replaceChildren();
  const line = document.createElement("p");
  line.className = "error-text";
  line.textContent = message;
  box.append(line);
}

function row(table, name, value) {
  const tr = document.createElement("tr");
  const key = document.createElement("th");
  key.textContent = name;
  const cell = document.createElement("td");
  cell.append(value);
  tr.append(key, cell);
  table.append(tr);
}

function renderResult(report) {
  $("result-card").hidden = false;
  const box = $("result");
  box.replaceChildren();

  const where = document.createElement("p");
  where.className = "note";
  where.append("Written to ");
  const code = document.createElement("code");
  code.textContent = report.directory;
  where.append(code);
  box.append(where);

  const table = document.createElement("table");
  table.className = "kv";
  row(table, "structure",
      `${report.n_atoms.toLocaleString()} atoms, ${report.n_grains} grains averaging ` +
      `${report.grain_size} Å across, in a ${report.box} Å cube`);
  row(table, "density",
      `${(100 * report.density).toFixed(1)} % of a perfect single crystal, ` +
      `closest pair ${report.min_separation} Å`);
  row(table, "builder", report.builder === "atomsk"
      ? "atomsk"
      : "the built in fallback, a few percent less dense at the boundaries than atomsk");
  const cite = document.createElement("a");
  cite.href = report.potential.url;
  cite.target = "_blank";
  cite.rel = "noreferrer";
  cite.textContent = report.potential.citation;
  row(table, "potential", cite);
  row(table, "run",
      `${report.run.atom_steps.toLocaleString()} atom-steps: roughly ` +
      `${report.run.minutes_four_cores.toFixed(0)} to ` +
      `${report.run.minutes_one_core.toFixed(0)} minutes`);
  box.append(table);

  const steps = document.createElement("ol");
  steps.className = "steps";
  const run = document.createElement("li");
  run.append("Run it:");
  run.append(pre(`cd ${report.directory}\nlmp -in in.compression`));
  const look = document.createElement("li");
  look.append("Then open the last frame here, or from the terminal:");
  look.append(pre(
    `ptmipf compression.dump --structures ${report.structure} --direction z \\\n` +
    `    --legend key.png --render map.png --hide-other`));
  steps.append(run, look);
  box.append(steps);

  const open = document.createElement("a");
  open.className = "button primary";
  open.href = "/?path=" + encodeURIComponent(report.relative_xyz);
  open.textContent = "Open the as-built structure in the analysis page";
  box.append(open);

  const files = document.createElement("p");
  files.className = "muted note";
  files.textContent = "Files: " + report.files.join(", ");
  box.append(files);
}

function pre(text) {
  const block = document.createElement("pre");
  block.textContent = text;
  return block;
}

async function init() {
  const saved = localStorage.getItem("ptmipf-theme");
  if (saved) document.documentElement.dataset.theme = saved;
  $("theme-toggle").addEventListener("click", () => {
    const root = document.documentElement;
    const dark = root.dataset.theme
      ? root.dataset.theme === "dark"
      : matchMedia("(prefers-color-scheme: dark)").matches;
    root.dataset.theme = dark ? "light" : "dark";
    localStorage.setItem("ptmipf-theme", root.dataset.theme);
  });

  try {
    state.meta = await api("/api/examples");
  } catch (error) {
    setStatus(error.message, "error");
    return;
  }
  if (!state.meta.atomsk) {
    const note = $("atomsk-note");
    note.textContent = state.meta.atomsk_help;
    note.hidden = false;
    $("builder").value = "voronoi";
  }
  renderCatalogue();
  for (const id of ["box", "grains", "strain", "strain-rate", "temperature", "seed"]) {
    $(id).addEventListener("input", renderCatalogue);
  }
  setStatus(`serving ${state.meta.root}`, "");
}

init().catch((error) => setStatus(error.message, "error"));

File: ptmipf/examples.py
"""Ready-made starter examples: build a polycrystal, fetch the potential, write the run.

One call produces a directory that contains everything needed to deform a small
polycrystal and then look at the result with ptm-ipf: the structure in both
LAMMPS and extended XYZ form, the EAM potential from NIST, the LAMMPS input,
the grain orientations that were built, and a README saying what to type.

The defaults are sized for a laptop.  Nothing here runs LAMMPS; it writes the
run and says what it will cost, because a tool that quietly starts a multi
minute job on someone's machine is a worse tool.
"""

from __future__ import annotations

import json
from dataclasses import asdict, dataclass
from pathlib import Path

__all__ = ["DEFAULTS", "ExampleSpec", "build_example", "main"]


@dataclass
class ExampleSpec:
    """Everything that decides what an example contains."""

    element: str = "Cu"
    box: float = 48.0  #: cube edge in angstrom
    n_grains: int = 6
    temperature: float = 300.0
    strain: float = 0.10
    strain_rate: float = 2.5e-3  #: engineering strain per picosecond
    seed: int = 1
    builder: str = "atomsk"

    def validate(self) -> None:
        from .potentials import potential_for

        potential_for(self.element)
        if not 12.0 <= self.box <= 200.0:
            raise ValueError("box must be between 12 and 200 angstrom")
        if not 1 <= self.n_grains <= 64:
            raise ValueError("n_grains must be between 1 and 64")
        if not 0.005 <= self.strain <= 0.5:
            raise ValueError("strain must be between 0.005 and 0.5")
        if not 1e-5 <= self.strain_rate <= 0.1:
            raise ValueError("strain rate must be between 1e-5 and 0.1 per picosecond")
        if not 1.0 <= self.temperature <= 2000.0:
            raise ValueError("temperature must be between 1 and 2000 K")


#: The catalogue the interface offers, chosen to run in a couple of minutes.
DEFAULTS = {
    "cu_compression": ExampleSpec(element="Cu", box=48.0, n_grains=6),
    "al_compression": ExampleSpec(element="Al", box=54.0, n_grains=6),
    "ni_compression": ExampleSpec(element="Ni", box=47.0, n_grains=6),
    "fe_compression": ExampleSpec(element="Fe", box=46.0, n_grains=6),
}


def example_directory(root, spec: ExampleSpec) -> Path:
    """Where an example lands: one directory per element under ``examples``."""
    name = f"{spec.element.lower()}_{spec.n_grains}grain_compression"
    return Path(root) / "examples" / name


def build_example(root, spec: ExampleSpec | None = None, **overrides) -> dict:
    """Build one example under ``<root>/examples/`` and report what was made.

    Returns a dictionary the web interface renders directly: the directory, the
    files, the structure statistics, the run cost and the commands to type.
    """
    from .lammps import compression_script
    from .polycrystal import build_polycrystal
    from .potentials import download_potential, potential_for

    spec = spec or ExampleSpec()
    if overrides:
        spec = ExampleSpec(**{**asdict(spec), **overrides})
    spec.validate()

    potential = potential_for(spec.element)
    directory = example_directory(root, spec)
    directory.mkdir(parents=True, exist_ok=True)

    # The download first: it is the only step that can fail for reasons outside
    # this machine, and failing before building saves the wait.
    potential_path = download_potential(spec.element, directory)

    crystal = build_polycrystal(
        spec.element,
        spec.box,
        spec.n_grains,
        directory,
        structure=potential.structure,
        a0=potential.a0,
        seed=spec.seed,
        builder=spec.builder,
    )

    script, settings = compression_script(
        potential,
        crystal.files["data"],
        crystal.n_atoms,
        temperature=spec.temperature,
        strain=spec.strain,
        strain_rate=spec.strain_rate,
        seed=spec.seed * 977 + 13,
    )
    (directory / "in.compression").write_text(script, encoding="utf-8", newline="\n")

    # The orientations that were built, so a map can be checked against them.
    (directory / "grains.json").write_text(
        json.dumps(
            {
                "element": crystal.element,
                "structure": crystal.structure,
                "a0": crystal.a0,
                "box": crystal.box,
                "builder": crystal.builder,
                "seeds": crystal.seeds.tolist(),
                "rotations": crystal.rotations.tolist(),
            },
            indent=1,
        ),
        encoding="utf-8",
        newline="\n",
    )

    report = {
        "directory": str(directory),
        "relative": str(Path("examples") / directory.name),
        "element": spec.element,
        "structure": crystal.structure,
        "builder": crystal.builder,
        "n_atoms": crystal.n_atoms,
        "n_grains": crystal.n_grains,
        "box": round(crystal.box, 3),
        "grain_size": round(crystal.mean_grain_size, 1),
        "density": round(crystal.density, 4),
        "min_separation": round(crystal.min_separation, 3),
        "potential": {
            "file": potential_path.name,
            "citation": potential.citation,
            "url": potential.entry_url,
            "pair_style": potential.pair_style,
        },
        "run": {k: (round(v, 6) if isinstance(v, float) else v) for k, v in settings.items()},
        "files": sorted(p.name for p in directory.iterdir() if p.is_file()),
        "spec": asdict(spec),
    }
    (directory / "README.md").write_text(_readme(report, crystal), encoding="utf-8", newline="\n")
    report["files"] = sorted(p.name for p in directory.iterdir() if p.is_file())
    return report


def _readme(report: dict, crystal) -> str:
    run = report["run"]
    low = run["minutes_four_cores"]
    high = run["minutes_one_core"]
    builder_note = (
        ""
        if report["builder"] == "atomsk"
        else "\nThis structure was built by the fallback NumPy builder, not atomsk, "
        "because atomsk was not found on this machine. It is a few percent less "
        "dense at the grain boundaries, which softens the elastic response a "
        "little. Install atomsk from https://atomsk.univ-lille.fr and rebuild for "
        "the reference structure.\n"
    )
    return f"""# {report["element"]} polycrystal under compression

{crystal.summary()}.
Closest pair as built: {report["min_separation"]} A, which the minimisation at the
top of the input script relaxes before any dynamics runs.
{builder_note}
## Run it

```bash
lmp -in in.compression            # or: mpirun -np 4 lmp -in in.compression
```

About {run["atom_steps"]:,} atom-steps: roughly {low:.0f} to {high:.0f} minutes
depending on how many cores it gets.

It writes `stress_strain.txt` (engineering strain against stress in GPa),
`compression.dump` ({run["load_steps"] // run["dump_every"]} frames) and
`compressed.lmp`.

## Look at it

```bash
ptmipf compression.dump --structures {report["structure"]} --direction z \\
    --legend key.png --render map.png --hide-other
ptmipf-ui compression.dump        # or open it in the web interface
```

The last frame is the deformed structure. `--frame 0` is the starting one, which
should be a clean set of randomly oriented grains: that is the control that says
the build and the colouring are both working.

## What is here

| file | what it is |
| --- | --- |
| `{crystal.files.get("data", "structure.lmp")}` | the structure, LAMMPS data format |
| `{crystal.files.get("xyz", "structure.xyz")}` | the same structure for ptm-ipf |
| `in.compression` | the LAMMPS input |
| `{report["potential"]["file"]}` | the EAM potential, from NIST |
| `grains.json` | the seed points and rotation matrices that were built |

The potential is {report["potential"]["citation"]}, taken from the NIST
Interatomic Potentials Repository, {report["potential"]["url"]}.
Cite it if you publish anything from this.
"""


def main(argv=None) -> int:
    """``ptmipf-example``: build a starter example from the terminal."""
    import argparse

    from .potentials import POTENTIALS

    parser = argparse.ArgumentParser(
        prog="ptmipf-example",
        description="Build a small polycrystal, fetch its EAM potential from NIST "
        "and write a LAMMPS compression run for it.",
    )
    parser.add_argument(
        "element", nargs="?", default="Cu", choices=sorted(POTENTIALS), help="default: Cu"
    )
    parser.add_argument("--root", default=".", help="examples land under <root>/examples")
    parser.add_argument("--box", type=float, help="cube edge in angstrom")
    parser.add_argument("--grains", type=int, help="number of grains")
    parser.add_argument("--strain", type=float, help="engineering strain to reach")
    parser.add_argument("--strain-rate", type=float, help="engineering strain per picosecond")
    parser.add_argument("--temperature", type=float, help="kelvin")
    parser.add_argument("--seed", type=int, help="random seed")
    parser.add_argument(
        "--builder",
        choices=("atomsk", "voronoi", "auto"),
        default="atomsk",
        help="atomsk builds at full density and is the default; voronoi is the "
        "built-in fallback; auto uses atomsk when it is installed",
    )
    args = parser.parse_args(argv)

    spec = DEFAULTS.get(f"{args.element.lower()}_compression", ExampleSpec(element=args.element))
    overrides = {"builder": args.builder}
    for name, value in (
        ("box", args.box),
        ("n_grains", args.grains),
        ("strain", args.strain),
        ("strain_rate", args.strain_rate),
        ("temperature", args.temperature),
        ("seed", args.seed),
    ):
        if value is not None:
            overrides[name] = value

    try:
        report = build_example(args.root, spec, **overrides)
    except (RuntimeError, ValueError) as exc:
        raise SystemExit(str(exc)) from None

    run = report["run"]
    print(f"wrote {report['directory']}")
    print(
        f"  {report['n_atoms']} atoms, {report['n_grains']} grains averaging "
        f"{report['grain_size']} A, {100 * report['density']:.1f} % dense "
        f"({report['builder']} builder)"
    )
    print(
        f"  {run['atom_steps']:,} atom-steps: roughly "
        f"{run['minutes_four_cores']:.0f} to {run['minutes_one_core']:.0f} minutes"
    )
    print(f"  cd {report['directory']} && lmp -in in.compression")
    return 0


if __name__ == "__main__":  # pragma: no cover
    raise SystemExit(main())
